File: Python/122.Best Time to Buy and Sell Stock II .py
#用一个数组表示股票每天的价格，数组的第i个数表示股票在第i天的价格。 如果允许多次交易(买一个再卖掉，再买)，
#但是，在买下一个前必须卖掉上一个，求最大的收益
class Solution(object):
    def maxProfit(self, prices):
        l  = len(prices)
        if l<=1:
            return 0
        maxpro = 0
        for i in range(1,l):
            if prices[i]>prices[i-1]:
                maxpro += (prices[i]-prices[i-1])
        return maxpro


s = Solution()
print(s.maxProfit([84,22,3,44,2]))

#只要后一天大于前一天，都算收益

File: C++/122.Best Time to Buy and Sell Stock II.cpp
/*
最佳的购买和抛售股票的时机，不限购买次数，求最大收益。
*/
/*
思路：找连续增加的数列即可。
 */

#include <iostream>
#include <vector>
#include <windows.h>
#include <windef.h>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	int length = prices.size();
        if(length==0)	return 0;
        int maxpro = 0;
        for(int i=0 ; i < length-1; ++i)
        {
        	int start = i;
        	while(i <length-1 && prices[i]<prices[i+1])	++i;
        	maxpro += prices[i]-prices[start];
        }
        return maxpro;
    }
};

int main()
{
	Solution s;
	vector<int> nums;
	nums.push_back(7);
	nums.push_back(1);
	nums.push_back(5);
	nums.push_back(3);
	nums.push_back(6);
	nums.push_back(8);
	//nums.push_back(2);
	cout<<s.maxProfit(nums)<<endl;
}